Remove the c-compile package

rneugeba opened the pull request:
With the new Alpine base image, we can now simply add the dev tools we need to compile the small bits of C code we have.

- Add compile tools as a stage to multi-stage builds
- Convert all affected packages to use the git tree hash
- Update packages which use the new packages
- Update all YAML files

Note: I have not updated the builds in `./projects` and some of them still use `c-compile`
Note: I have not pushed new the kernel builds. They will pick up the new toybox soon enough

fix #1731

justincormack commented on pkg/init/Dockerfile:
```diff
@@ -1,9 +1,20 @@
-FROM alpine:3.5
+FROM linuxkit/alpine:f0169b60fb260d74025496ae6fd93213fecaba8f@sha256:23743c7206ebe8a609442c5ac7084a26ed45ce8f5213960428bca264225849f1 AS build
+# XXX There is an issue with the bsd-compat-headers package I don't
+# understand. Installing the file directly is yuk but works
```
can you see if this is fixed by https://github.com/linuxkit/linuxkit/pull/1781

rneugeba replied:
I build the new alpine base locally with `make tag` and then used the new image:
```FROM linuxkit/alpine:a09caa34e52e873b23114cd1d691a9a61ee1a10f AS build
RUN apk add \
    bsd-compat-headers \
    gcc \
    linux-headers \
    musl-dev
COPY ./start-stop-daemon.c .
```
and still get the error:
```
[no-c *]rn-mbp:t/linuxkit/pkg/init: make tag
docker build --no-cache --network=none -t linuxkit/init:4c7e24873e2239c04ce2cca2fb4ddfcc173be459 .
Sending build context to Docker daemon  43.01kB
Step 1/10 : FROM linuxkit/alpine:a09caa34e52e873b23114cd1d691a9a61ee1a10f AS build
 ---> ee89320575c0
Step 2/10 : RUN apk add     bsd-compat-headers     gcc     linux-headers     musl-dev
 ---> Running in 53659afe41e3
ERROR: bsd-compat-headers-0.7.1-r0: package mentioned in index not found (try 'apk update')
(1/15) Installing bsd-compat-headers (0.7.1-r0)
(2/15) Installing binutils-libs (2.28-r1)
(3/15) Installing binutils (2.28-r1)
(4/15) Installing gmp (6.1.2-r0)
(5/15) Installing isl (0.17.1-r0)
(6/15) Installing libgomp (6.3.0-r4)
(7/15) Installing libatomic (6.3.0-r4)
(8/15) Installing pkgconf (1.3.5-r0)
(9/15) Installing libgcc (6.3.0-r4)
(10/15) Installing mpfr3 (3.1.5-r0)
(11/15) Installing mpc1 (1.0.3-r0)
(12/15) Installing libstdc++ (6.3.0-r4)
(13/15) Installing gcc (6.3.0-r4)
(14/15) Installing linux-headers (4.4.6-r2)
(15/15) Installing musl-dev (1.1.16-r8)
Executing busybox-1.26.2-r2.trigger
1 errors; 103 MiB in 25 packages
```

justincormack replied:
typo, hold on will try again

justincormack commented on pkg/init/Dockerfile:
```diff
@@ -1,9 +1,20 @@
+RUN cc -static -O2 -Wall -Werror -o start-stop-daemon start-stop-daemon.c
+
+# Pull in the latest 3.5 so we don't have to update/upgrade
+FROM alpine:3.5@sha256:58e1a1bb75db1b5a24a462dd5e2915277ea06438c3f105138f97eb53149673c4
```
this should use `edge` as that is what the linuxkit/alpine package is built from now (we may move to 3.6 when it is out).

rneugeba commented:
crap, I

rneugeba commented:
crap, I accidentally deleted the branch. restored, now

justincormack commented:
ok try again now with `linuxkit/alpine:c5021a113803f7608e3908014d316c3490183d0b` seems to be working now

rneugeba commented:
Thanks, that seems to work

rneugeba commented:
Once #1781 is merged, I'll rework this

justincormack commented:
Its merged now

rneugeba commented:
Pushed an update to use the update alpine base package
- fixed up existing users of the older alpine base package
- fixed up the `init` package to use `edge`
- updated all the YAML files which used older refs

justincormack commented:
Seems to have a rebase issue

rneugeba commented:
rebased

justincormack commented on pkg/ca-certificates/Dockerfile:
```diff
@@ -1,4 +1,4 @@
-FROM linuxkit/alpine:f0169b60fb260d74025496ae6fd93213fecaba8f as alpine
+FROM linuxkit/alpine:c5021a113803f7608e3908014d316c3490183d0b@sha256:4c7a80173c71ca5019dc56f40442b3a8345f141dd46593bd1fe6d130294d07cb as alpine
```
The plan was not to have sha256 hashes, as we are going to use notary so tags are fine.

rneugeba replied:
Yes, but don't have a lot of packages signed yet. I suggest we get rid of the `sha256` hashes once all packages/tools are signed

justincormack replied:
ok

justincormack commented on pkg/init/Dockerfile:
```diff
@@ -1,9 +1,17 @@
-  apk --no-cache upgrade -a && \
-  apk --no-cache add \
-  && rm -rf /var/cache/apk/*
+RUN cc -static -O2 -Wall -Werror -o start-stop-daemon start-stop-daemon.c
```
actually this version, which uses `containerd` no longer uses `start-stop-daemon` may as well just remove it.

justincormack replied:
can do this later if you like

rneugeba replied:
I can do that now. `linuxkit/init` is relatively easy as it has no further dependencies. `linuxkit/tini` is a pain as it triggers a chain of update.

rneugeba replied:
removed start-stop-daemon

justincormack commented on pkg/init/Dockerfile:
```diff
@@ -1,9 +1,6 @@
-  && rm -rf /var/cache/apk/*
-
+# Use sha256 here to get a fixed version
+FROM alpine:edge@sha256:99588bc8883c955c157d18fc3eaa4a3c1400c223e6c7cabca5f600a3e9f8d5cd
```
hmm, not ideal as this will be out of date (new `edge` images are not pushed very often, linuxkit/alpine does an upgrade). On the other hand, will remove `alpine` from here soon so lets leave this for now.

rneugeba replied:
yes, but otherwise we can't use the git tree hash...but as you said, this will be gone soon

justincormack replied:
well I was thinking you could use `linuxkit/alpine`, remove the mirror and then squash it

rneugeba replied:
that doesn't work because squashing only works with newly created layers. We could start with `linuxkit/alpine` then create a tar ball for a new root fs from that, and in a new `scratch` layer untar that...

justincormack commented:
Added a note that it fixes #1731
